org.osgi.test.junit5.cm/src/main/java/org/osgi/test/junit5/cm/ConfigurationExtension.java

```diff
@@ -0,0 +1,200 @@
+			if (configurationAdmin == null) {
+				throw new IllegalStateException("The configuration admin service cannot be found.");
+			}
+
+			Configuration configuration = null;
+
+			switch (configurationAnnotation.action()) {
+				case GET :
+					if (configurationAnnotation.factory()) {
+						configuration = ConfigAdminUtil.getConfigsByFactoryServicePid(configurationAdmin,
+							configurationAnnotation.pid(), configurationAnnotation.factoryName());
+					} else {
+						configuration = ConfigAdminUtil.getConfigsByServicePid(configurationAdmin,
+							configurationAnnotation.pid());
+					}
+					break;
+				case GET_OR_CREATE :
+					if (configurationAnnotation.factory()) {
+						configuration = configurationAdmin.getFactoryConfiguration(configurationAnnotation.pid(),
+							configurationAnnotation.factoryName());
+					} else {
+						configuration = configurationAdmin.getConfiguration(configurationAnnotation.pid());
+					}
+					break;
+
+				case CREATE :
+					if (configurationAnnotation.factory()) {
+						configuration = configurationAdmin.createFactoryConfiguration(configurationAnnotation.pid(),
+							null);
+					} else {
+						configuration = configurationAdmin.createFactoryConfiguration(configurationAnnotation.pid(),
+							null);
+					}
+					break;
+
+			}
+
+			if (configurationAnnotation.properties() != null) {
+				Map<String, ?> map = Stream.of(configurationAnnotation.properties())
+					.collect(Collectors.toMap((e) -> e.key(), (e) -> e.value()));
+				configuration.update(Dictionaries.asDictionary(map));
+			}
+			return configuration;
+
+		} catch (Exception e) {
+			throw new ExtensionConfigurationException(e.getMessage(), e);
+		} finally {
+			if (bundleContext != null && sref != null) {
+				bundleContext.ungetService(sref);
+			}
+		}
+	}
+
+	static void assertValidFieldCandidate(Field field) {
+		if (Modifier.isFinal(field.getModifiers()) || Modifier.isPrivate(field.getModifiers())
+			|| Modifier.isStatic(field.getModifiers())) {
+			throw new ExtensionConfigurationException("@" + InjectConfiguration.class.getSimpleName() + " field ["
+				+ field.getName() + "] must not be final, private or static.");
+		}
+	}
+
+	@Override
+	public Object resolveParameter(ParameterContext parameterContext, ExtensionContext extensionContext)
+		throws ParameterResolutionException {
+
+		BundleContext bundleContext = FrameworkUtil.getBundle(extensionContext.getRequiredTestClass())
+			.getBundleContext();
+
+		Optional<InjectConfiguration> configAnnotation = parameterContext.findAnnotation(InjectConfiguration.class);
+		Parameter parameter = parameterContext.getParameter();
+
+		return resolveReturnValue(configAnnotation.get(), bundleContext, extensionContext);
+	}
+
+	@Override
+	public boolean supportsParameter(ParameterContext parameterContext, ExtensionContext extensionContext)
+		throws ParameterResolutionException {
+
+		if (parameterContext.isAnnotated(InjectConfiguration.class)) {
+			return true;
+		}
```

Comment:
I'd still like to see some error checking here. If a field is annotated with `@InjectConfiguration` (or whatever the final name ends up being) but the type of the field is not `Configuration`, then we should throw an exception here.

Reply:
Ok,
If the `@InjectConfiguration` is set on Dictionaries or Maps, should we Inject  the properties of the configuration?

Reply:
If that's not too hard to do, it might be useful. Both would be useful, but Map would probably be the more useful of the two because it wouldn't be too hard to inject a `Configuration` object and then simply call `Configuration.getProperties()`.

Reply:
done